src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/DeclareCommands.java

```diff
@@ -0,0 +1,479 @@
+	// Not sure if we will need these..?
+	public static final int ARG_TYPE_RAWTEXT = 34;
+	public static final int ARG_TYPE_JSON = 37;
+	public static final int ARG_TYPE_COMMAND = 44;
+
+	public static final int ARG_FLAG_VALID = 0x100000;
+	public static final int ARG_FLAG_LITERAL = 0x200000;
+
+	private static Map<String, Consumer<ByteBuf>> argTypeReaders = new HashMap<>();
+	private static Map<String, Integer> peArgTypeRemaps = new HashMap<>();
+
+	static {
+		registerArgTypeReaders();
+		registerPEArgTypeRemaps();
+	}
+
+	private CommandNode[] allNodes;
+	private CommandNode[] topLevelNodes; // A subset of allNodes containing the starting "command" nodes
+	private int rootNodeIndex;
+
+	private static class CommandNode {
+		private final String name;
+		private final String argType;
+		private final String suggestion;
+		private final int[] children;
+		private final int redirect;
+		private final boolean isPathEnd;
+
+		public CommandNode(String name, String argType, String suggestion, int[] children, int redirect, boolean isPathEnd) {
+			this.name = name;
+			this.argType = argType;
+			this.suggestion = suggestion;
+			this.children = children;
+			this.redirect = redirect;
+			this.isPathEnd = isPathEnd;
+		}
+
+		@Override
+		public String toString() {
+			return Utils.toStringAllFields(this);
+		}
+
+		public boolean isLeaf() {
+			return this.getChildren().length == 0;
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public String getArgType() {
+			return argType;
+		}
+
+		public int[] getChildren() {
+			return children;
+		}
+	}
+
+	private static class PEArgumentNode {
+		private final String name;
+		private final String argType;
+		private final int nameIndex;
+
+		public PEArgumentNode(CommandNode node, PECommandsStructure peStruct) {
+			this.name = node.getName();
+			this.argType = node.getArgType();
+
+			// Cache literal index, if this is a literal argument
+			if (this.argType == null) {
+				this.nameIndex = peStruct.registerLiteral(this.name);
+			} else {
+				this.nameIndex = -1;
+			}
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public String getArgType() {
+			return argType;
+		}
+
+		public int getNameIndex() {
+			return nameIndex;
+		}
+	}
+
+	private static class PECommandsStructure {
+		private Map<CommandNode, List<List<PEArgumentNode>>> overloadsRegistry = new HashMap<>();
+		private LinkedHashMap<String, Integer> literalRegistry = new LinkedHashMap<>();
+
+		/**
+		 * Register a list of overloads for a top-level node.
+		 */
+		public void registerOverloads(CommandNode node, List<List<PEArgumentNode>> overloads) {
+			overloadsRegistry.put(node, overloads);
+		}
+
+		public List<List<PEArgumentNode>> getOverloads(CommandNode node) {
+			return overloadsRegistry.get(node);
+		}
+
+		/**
+		 * Register a literal in the internal literal registry if it was new, and return its index value.
+		 */
+		public int registerLiteral(String literal) {
```

Comment:
computeIfAbsent

Reply:
Thanks. Last time I programmed Java seriously was prior to lambdas, so I'm still catching up on all new functionality in the core classes.